### PR TITLE
Selective client

### PR DESCRIPTION
**Pull request by rrati**
Allow using unleash to control which client is used by specific clusters

### CODE REVIEW

**Comment by rrati** [comment]
@rporres This is for you

depends on #1400

File: reconcile/utils/oc.py [hunk 1]
```diff
@@ -620,6 +635,245 @@ def _run_json(self, cmd, allow_not_found=False):
...
+        k, group_version = self._parse_kind(kind)
+        obj_client = self._get_obj_client(group_version=group_version, kind=k)
+        try:
+            obj = obj_client.get(name=name, namespace=namespace)
+            return obj.to_dict()
+        except NotFoundError as e:
+            if allow_not_found:
+                return {}
+            else:
+                raise StatusCodeError(f"[{self.server}]: {e}")
```
Comment (Piojo) [inline reconcile/utils/oc.py:838]: Does `e` mention `name` and `namespace`? Unless they come from a vault, they'd be very useful to have when troubleshooting.

File: reconcile/utils/oc.py [hunk 2]
```diff
@@ -620,6 +635,245 @@ def _run_json(self, cmd, allow_not_found=False):
...
+            # this is a new kind so add it
+            kgv[kind] = [new]
+        else:
+            # this kind already exists, so check if this apigroup has already
+            # been added as an option.  If this apigroup is the preferred one,
+            # then replace the apigroup/version so that the preferred
+            # apigroup/version is used instead of a non-preferred one
+            group = new.split('/', 1)[0]
+            new_group = True
+            for pos in range(len(kgv[kind])):
```
Comment (Piojo) [inline reconcile/utils/oc.py:864]: Wouldn't this be easier to read?
```python
for pos, item in enumerate(kgv[kind]):
    if group in item:
         new_group = False
         if preferred:
              kgv[kind][pos] = new
```

File: reconcile/utils/oc.py [hunk 3]
```diff
@@ -620,6 +635,245 @@ def _run_json(self, cmd, allow_not_found=False):
...
+                    for kind in r.keys():
+                        if len(apigroup) == 0:
+                            gv = f'{version}'
+                        else:
+                            gv = f'{apigroup}/{version}'
+                        kind_groupversion = self.add_group_kind(
+                            kind, kind_groupversion, gv, obj.preferred)
+        return kind_groupversion
+
+    def get_items(self, kind, **kwargs):
```
[the remark this answers is not in the document]
Comment (rrati) [reply reconcile/utils/oc.py:777]: I do not plan to change this on this iteration.  The API definitions are copied from the deprecated, and as such I do not want to change anything more than needed.   The big change here is to use native python client, and as such I want to minimize other changes not related to that no matter how much better they may be

File: reconcile/utils/oc.py [hunk 4]
```diff
@@ -620,6 +635,245 @@ def _run_json(self, cmd, allow_not_found=False):
...
+        obj_client = self._get_obj_client(group_version=group_version, kind=k)
+
+        namespace = ''
+        if 'namespace' in kwargs:
+            namespace = kwargs['namespace']
+            # for cluster scoped integrations
+            # currently only openshift-clusterrolebindings
+            if namespace != 'cluster':
+                if not self.project_exists(namespace):
+                    return []
```
[the remark this answers is not in the document]
Comment (rrati) [reply reconcile/utils/oc.py:791]: See above

File: reconcile/utils/oc.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmelis) [reply reconcile/utils/oc.py:794]: +1 to not fixing this in this PR. We have this all over the codebase.  But, I do agree with changing our style to stop doing this.

File: reconcile/utils/oc.py (hunk 1 above)
Comment (jmelis) [reply reconcile/utils/oc.py:855]: +1, very good point. This goes in #1400 though.

Comment (rrati) [reply reconcile/utils/oc.py:855]: ATM, this is the same error message raised by the original get function.  I would prefer not to change that here and do it in a separate PR once this is merged.  My intention was to keep inputs/outpus the same.

File: reconcile/utils/oc.py [hunk 5]
```diff
@@ -957,8 +959,25 @@ class OC:
     def __new__(cls, cluster_name, server, token, jh=None, settings=None,
                 init_projects=False, init_api_resources=False,
                 local=False):
-        use_native = os.environ.get('USE_NATIVE_CLIENT', '').lower() \
-            in ['true', 'yes']
+        use_native = os.environ.get('USE_NATIVE_CLIENT', '')
```
Comment (BumbleFeng) [inline reconcile/utils/oc.py:962]: current USE_NATIVE_CLIENT will be either "false" or "true", but in both conditions, if use_native will always be true.
```
>>> bool("false")
True
```

Comment (rrati) [reply reconcile/utils/oc.py:962]: Fixed

**Review by BumbleFeng — APPROVED** [review]

**Pull request merged**